File: stacks/net/kuma-provision.js
#!/usr/bin/env node
/*
 * Provision the ladybird monitor set in Uptime Kuma.
 *
 * Credentials come from the environment, never from this file:
 *   KUMA_USER  (default: admin)
 *   KUMA_PW    (required)
 *   KUMA_URL   (default: http://127.0.0.1:3001)
 *
 * Prints a plan and exits. Pass --apply to actually create anything.
 * Idempotent: monitors are matched by name, existing ones are left alone.
 */
const io = require("socket.io-client");
const crypto = require("crypto");

const URL = process.env.KUMA_URL || "http://127.0.0.1:3001";
const USER = process.env.KUMA_USER || "admin";
const PW = process.env.KUMA_PW;
const APPLY = process.argv.includes("--apply");
const HOST = process.env.LADYBIRD_HOST || "192.168.0.13";

if (!PW) { console.error("KUMA_PW is not set. Refusing to continue."); process.exit(1); }

const push = () => crypto.randomBytes(16).toString("hex");
const PUSH_TOKENS = { "frigate-recording": push(), "storage-guard": push() };

// Sensible defaults: 60s beat, 3 retries at 60s => ~3 min before it pages,
// so a two-second blip stays quiet.
const base = {
  interval: 60, retryInterval: 60, resendInterval: 0, maxretries: 3,
  timeout: 20, accepted_statuscodes: ["200-299"], active: true,
  method: "GET", maxredirects: 10, ignoreTls: false, upsideDown: false,
  expiryNotification: false, description: null,
};

const STATS = `http://${HOST}:5000/api/stats`;

// parent is a NAME here; resolved to an id before sending.
const MONITORS = [
  { name: "frigate-api", type: "keyword", url: `http://${HOST}:5000/api/version`,
    keyword: "0.17", invertKeyword: false,
    description: "Frigate API answering. Parent of every frigate-* check." },

  { name: "frigate-driveway-stream", type: "json-query", url: STATS, parent: "frigate-api",
    jsonPath: "cameras.driveway.camera_fps > 0", expectedValue: "true",
    description: "RTSP pull alive for driveway." },
  { name: "frigate-backyard-stream", type: "json-query", url: STATS, parent: "frigate-api",
    jsonPath: "cameras.backyard.camera_fps > 0", expectedValue: "true",
    description: "RTSP pull alive for backyard." },

  // These two would have caught detect.enabled defaulting to false in 0.17.
  { name: "frigate-driveway-detect", type: "json-query", url: STATS, parent: "frigate-api",
    jsonPath: "cameras.driveway.detection_enabled", expectedValue: "true",
    description: "Object detection actually running on driveway." },
  { name: "frigate-backyard-detect", type: "json-query", url: STATS, parent: "frigate-api",
    jsonPath: "cameras.backyard.detection_enabled", expectedValue: "true",
    description: "Object detection actually running on backyard." },

  { name: "frigate-ui", type: "port", hostname: HOST, port: 8971, parent: "frigate-api",
    description: "Authenticated Frigate UI." },

  // Fed by frigate-guard.sh, which only beats when streams, detection AND
  // recording-to-disk are all healthy. Silence is the alert.
  { name: "frigate-recording", type: "push", parent: "frigate-api",
    interval: 300, maxretries: 1, pushToken: PUSH_TOKENS["frigate-recording"],
    description: "Recording segments hitting disk. Beat by frigate-guard.timer (2 min)." },

  // Cameras are NOT children of frigate-api: a camera can die while Frigate is fine.
  { name: "cam-driveway", type: "ping", hostname: "10.10.10.201", maxretries: 4,
    description: "Camera 1 on the isolated island." },
  { name: "cam-backyard", type: "ping", hostname: "10.10.10.202", maxretries: 4,
    description: "Camera 2 on the isolated island." },

  { name: "home-assistant", type: "http", url: `http://${HOST}:8123`,
    accepted_statuscodes: ["200-299", "300-399"] },
  { name: "mosquitto", type: "port", hostname: HOST, port: 1883 },
  { name: "jellyfin", type: "http", url: `http://${HOST}:8096`,
    accepted_statuscodes: ["200-299", "300-399"] },

  { name: "storage-guard", type: "push", interval: 900, maxretries: 1,
    pushToken: PUSH_TOKENS["storage-guard"],
    description: "Beat by disk-guard.timer (10 min) only while storage is healthy." },

  { name: "wan", type: "ping", hostname: "1.1.1.1", maxretries: 5,
    description: "Internet reachability. Independent of everything else." },
];

const socket = io(URL, { transports: ["websocket"], reconnection: false });
const emit = (ev, ...a) => new Promise((res, rej) => {
  const t = setTimeout(() => rej(new Error(`timeout on ${ev}`)), 20000);
  socket.emit(ev, ...a, (r) => { clearTimeout(t); r && r.ok === false ? rej(new Error(r.msg || ev)) : res(r); });
});

let monitorList = {}, notificationList = [];
socket.on("monitorList", (l) => { monitorList = l || {}; });
socket.on("notificationList", (l) => { notificationList = l || []; });

socket.on("connect_error", (e) => { console.error("connect failed:", e.message); process.exit(1); });

socket.on("connect", async () => {
  try {
    await emit("login", { username: USER, password: PW, token: "" });
    console.log(`Logged in to ${URL} as ${USER}\n`);
    await new Promise((r) => setTimeout(r, 1500)); // let pushed lists arrive

    const existing = new Set(Object.values(monitorList).map((m) => m.name));
    const notifIDs = {};
    notificationList.forEach((n) => { notifIDs[n.id] = true; });

    if (notificationList.length === 0) {
      console.log("!! NO NOTIFICATION CHANNELS EXIST.");
      console.log("!! Monitors will be created but CANNOT alert anyone.");
      console.log("!! Create one in Settings > Notifications, then re-run to attach it.\n");
    } else {
      console.log(`Attaching ${notificationList.length} notification channel(s): ` +
        notificationList.map((n) => n.name).join(", ") + "\n");
    }

    const ids = {};
    Object.values(monitorList).forEach((m) => { ids[m.name] = m.id; });

    for (const spec of MONITORS) {
      if (existing.has(spec.name)) {
        console.log(`  skip   ${spec.name} (already exists)`);
        continue;
      }
      const { parent, ...rest } = spec;
      const mon = { ...base, ...rest, notificationIDList: notifIDs,
                    parent: parent ? ids[parent] ?? null : null };

      if (!APPLY) {
        console.log(`  PLAN   ${spec.name}  [${spec.type}]` + (parent ? ` -> child of ${parent}` : ""));
        continue;
      }
      const r = await emit("add", mon);
      ids[spec.name] = r.monitorID;
      console.log(`  create ${spec.name}  [${spec.type}] id=${r.monitorID}` +
                  (parent ? ` -> child of ${parent}` : ""));
    }

    if (APPLY) {
      console.log("\nPush URLs - write these to the server, then chmod 600:");
      for (const [n, t] of Object.entries(PUSH_TOKENS)) {
        const f = n === "storage-guard" ? "kuma-push-url.txt" : "kuma-frigate-push-url.txt";
        console.log(`  /opt/stacks/net/${f}\n    ${URL}/api/push/${t}?`);
      }
    } else {
      console.log("\nPlan only. Re-run with --apply to create these.");
    }
    socket.close(); process.exit(0);
  } catch (e) {
    console.error("FAILED:", e.message); socket.close(); process.exit(1);
  }
});

File: stacks/net/kuma-provision.py
#!/usr/bin/env python3
"""Apply kuma-monitors.yml to a running Uptime Kuma.

The YAML stays the source of truth; this pushes it into Kuma's socket.io API so
the monitors do not have to be clicked in by hand. Idempotent: monitors are
matched by name, created when missing, updated when they have drifted.

Read docs/uptime-kuma.md before running. In particular this script deliberately
does NOT touch Settings (base URL, timezone, retention) - that call replaces the
whole settings object, and getting it wrong is a worse outage than clicking
three fields once.

This is an API client, not a plugin - run it anywhere that can reach port 3001.
NOT inside the Kuma container: that is a Node image, and anything installed in
it is lost on the next `docker compose pull`.

On the host, Debian 13 marks its Python externally-managed (PEP 668), so a bare
`pip install` refuses. Use a venv:

    sudo apt install -y python3-venv
    python3 -m venv /opt/stacks/net/.venv
    /opt/stacks/net/.venv/bin/pip install uptime-kuma-api pyyaml

    export KUMA_PASSWORD='...'          # never pass this as an argument - argv is world-readable in ps
    export NTFY_TOPIC='ladybird-xxxxx'  # optional; creates + attaches the ntfy notification

    /opt/stacks/net/.venv/bin/python /opt/stacks/net/kuma-provision.py           # dry run: changes nothing
    /opt/stacks/net/.venv/bin/python /opt/stacks/net/kuma-provision.py --apply   # actually writes

Environment:
    KUMA_URL       default http://192.168.0.13:3001
    KUMA_USERNAME  default admin
    KUMA_PASSWORD  required
    NTFY_TOPIC     optional
"""

import os
import sys
from pathlib import Path

try:
    import yaml
    from uptime_kuma_api import UptimeKumaApi, MonitorType, NotificationType
except ImportError as exc:
    sys.exit(f"missing dependency: {exc}\n  pip install uptime-kuma-api pyyaml")

DEFINITION = Path(__file__).with_name("kuma-monitors.yml")

# The YAML's type names, kept short and readable there, mapped to Kuma's.
# JSON_QUERY arrived in Kuma 1.23; fall back to the raw wire value so an older
# client library does not break the whole run.
TYPES = {
    "http": MonitorType.HTTP,
    "keyword": MonitorType.KEYWORD,
    "json-query": getattr(MonitorType, "JSON_QUERY", "json-query"),
    "port": MonitorType.PORT,
    "ping": MonitorType.PING,
    "push": MonitorType.PUSH,
}


def build_kwargs(m, parent_id, notification_ids):
    """Translate one YAML monitor block into Kuma's field names."""
    kw = {
        "name": m["name"],
        "interval": m.get("interval", m.get("heartbeat_interval", 60)),
        "maxretries": m.get("retries", 0),
        "retryInterval": m.get("retry_interval", 60),
    }
    if parent_id is not None:
        kw["parent"] = parent_id
    if notification_ids:
        kw["notificationIDList"] = {str(i): True for i in notification_ids}

    for src, dst in (
        ("url", "url"),
        ("hostname", "hostname"),
        ("port", "port"),
        ("keyword", "keyword"),
        ("json_path", "jsonPath"),
        ("expected_value", "expectedValue"),
        ("ignore_tls", "ignoreTls"),
        ("expiry_notification", "expiryNotification"),
    ):
        if src in m:
            kw[dst] = m[src]

    if "accepted_status" in m:
        kw["accepted_statuscodes"] = [str(c) for c in m["accepted_status"]]

    return kw


def main():
    apply = "--apply" in sys.argv
    url = os.environ.get("KUMA_URL", "http://192.168.0.13:3001")
    username = os.environ.get("KUMA_USERNAME", "admin")
    password = os.environ.get("KUMA_PASSWORD")
    ntfy_topic = os.environ.get("NTFY_TOPIC")

    if not password:
        sys.exit("KUMA_PASSWORD is not set. Export it rather than passing it as an argument.")

    spec = yaml.safe_load(DEFINITION.read_text())
    if not apply:
        print("DRY RUN - nothing will be written. Re-run with --apply.\n")

    api = UptimeKumaApi(url)
    api.login(username, password)
    try:
        existing = {mon["name"]: mon for mon in api.get_monitors()}

        # ---- notification -------------------------------------------------
        notification_ids = []
        if ntfy_topic:
            found = next((n for n in api.get_notifications() if n["name"] == "ntfy"), None)
            if found:
                notification_ids = [found["id"]]
                print("notification  ntfy: exists")
            elif apply:
                res = api.add_notification(
                    name="ntfy",
                    type=NotificationType.NTFY,
                    isDefault=True,
                    applyExisting=True,
                    ntfyserverurl=spec["notifications"][0]["server"],
                    ntfytopic=ntfy_topic,
                    ntfyPriority=spec["notifications"][0]["priority"],
                    ntfyAuthenticationMethod="none",
                )
                notification_ids = [res["id"]]
                print("notification  ntfy: created - send a test from the UI before trusting it")
            else:
                print("notification  ntfy: would create")
        else:
            print("notification  ntfy: skipped (NTFY_TOPIC unset) - monitors will alert nowhere")

        # ---- groups first, so children have a parent to point at ----------
        group_ids = {}
        for g in spec.get("groups", []):
            if g["name"] in existing:
                group_ids[g["name"]] = existing[g["name"]]["id"]
                print(f"group         {g['name']}: exists")
            elif apply:
                res = api.add_monitor(type=MonitorType.GROUP, name=g["name"])
                group_ids[g["name"]] = res["monitorID"]
                print(f"group         {g['name']}: created")
            else:
                print(f"group         {g['name']}: would create")

        # ---- monitors ------------------------------------------------------
        for m in spec["monitors"]:
            kw = build_kwargs(m, group_ids.get(m.get("group")), notification_ids)
            name = m["name"]
            if name in existing:
                if apply:
                    api.edit_monitor(existing[name]["id"], **kw)
                    print(f"monitor       {name}: updated")
                else:
                    print(f"monitor       {name}: would update")
            elif apply:
                api.add_monitor(type=TYPES[m["type"]], **kw)
                print(f"monitor       {name}: created")
            else:
                print(f"monitor       {name}: would create ({m['type']})")

        # ---- the push URL the storage guard needs --------------------------
        if apply:
            guard = next((mon for mon in api.get_monitors() if mon["name"] == "Storage guard"), None)
            if guard and guard.get("pushToken"):
                print(
                    f"\nStorage guard push URL:\n  {url}/api/push/{guard['pushToken']}\n\n"
                    "Put it on the server, then beat it once:\n"
                    "  printf '%s\\n' '<url above>' > /opt/stacks/net/kuma-push-url.txt\n"
                    "  chmod 600 /opt/stacks/net/kuma-push-url.txt\n"
                    "  /opt/stacks/net/disk-guard.sh"
                )
    finally:
        api.disconnect()


if __name__ == "__main__":
    main()
